journal/src/main/scala/com/evolutiongaming/kafka/journal/util/ResourcePool.scala

```diff
@@ -0,0 +1,276 @@
+package com.evolutiongaming.kafka.journal.util
+
+import cats.instances.all._
+import cats.effect.implicits._
+import cats.effect.kernel.{Async, Deferred, Sync}
+import cats.effect.std.Mutex
+import cats.effect.{Ref, Resource}
+import cats.syntax.all._
+import com.evolutiongaming.catshelper.Log
+
+import java.util.UUID
+import java.util.concurrent.TimeoutException
+import scala.concurrent.duration.FiniteDuration
+
+private[journal] trait ResourcePool[F[_], O] {
+
+  /**
+   * `acquire` function of the Resource semantically blocks if there are no objects available.
+   *
+   * `release` function of the Resource returns the object to the pool.
+   * Note that it will NOT deallocate the object.
+   */
+  def borrow: Resource[F, O]
+}
+
+private[journal] object ResourcePool {
+
+  type ObjectId = UUID
+  type WaiterId = UUID
+  type Waiter[F[_], O] = Deferred[F, Either[Throwable, (ObjectId, O)]]
+
+  case class AllocatedObject[F[_], O](value: O, deallocate: F[Unit], state: AllocatedObject.State)
+
+  object AllocatedObject {
+    sealed abstract class State(val isFree: Boolean)
+
+    object State {
+      case object Used extends State(false)
+      case class Free(releasedAt: FiniteDuration) extends State(true)
+    }
+  }
+
+  case class State[F[_], O](
+    objects: Map[ObjectId, AllocatedObject[F, O]],
+    waiters: Map[WaiterId, Waiter[F, O]],
+  ) {
+
+    def findFreeObject: Option[(ObjectId, AllocatedObject[F, O])] =
+      objects.collectFirst {
+        case (id, obj) if obj.state.isFree => (id, obj)
+      }
+
+    def markObjectAsUsed(id: ObjectId, obj: AllocatedObject[F, O]): State[F, O] =
+      State(
+        objects + (id -> obj.copy(state = AllocatedObject.State.Used)),
+        waiters
+      )
+
+    def markObjectAsFree(id: ObjectId, obj: AllocatedObject[F, O], releasedAt: FiniteDuration): State[F, O] =
+      State(
+        objects + (id -> obj.copy(state = AllocatedObject.State.Free(releasedAt))),
+        waiters
+      )
+
+    def addObject(id: ObjectId, obj: AllocatedObject[F, O]): State[F, O] =
+      State(
+        objects + (id -> obj),
+        waiters
+      )
+
+    def removeObject(id: ObjectId): State[F, O] =
+      State(
+        objects - id,
+        waiters
+      )
+
+    def addWaiter(id: WaiterId, waiter: Waiter[F, O]): State[F, O] =
+      State(
+        objects,
+        waiters + (id -> waiter),
+      )
+
+    def removeWaiter(id: WaiterId): State[F, O] =
+      State(
+        objects,
+        waiters - id
+      )
+  }
+
+  def fixedSize[F[_]: Async, O](
+    poolSize: Int,
+    acquireTimeout: FiniteDuration,
```

Comment:
why do we need this?

Reply:
I guess what you're suggesting is to implement timeout on client side and handle fiber cancelation on pool side?

Reply:
I'm not sure it's convenient to add `.timeout` on the caller side. If we want to support fiber cancellation we'll need to support it for both CE2 and CE3. I'm not familiar with how cancellation works in CE2 but I've glanced through the docs and it seems to be different from CE3. `ResourcePool` is private api and it's only used once, so I don't understand why we need this maintenance burden. My `acquireTimeout` implementation may be a crutch but it allows us to not think about CE2/CE3 compatibility.

Reply:
Do you want to say that passing `timeout` as argument is more "convenient" comparing to just using already provided api of `.timeout` ?